docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+| `ce`   | Official CloudEvents attributes defined in the spec, plus extension attributes if requested. |
+| `data` | Fields parsed from the event data, if requested. |
+
+Official attributes and extension attributes are both available to the
+expression under the prefix `ce`. This co-mingling is intended to ease the
+promotion of extension attributes to official attributes. The prefix `ce` was
+chosen for succintness and its suggestion that CloudEvents is the operative
+specification.
+
+Data fields are available to the expression under the prefix `data`.
+
+Parsing dynamic extension attributes and data fields may be costly. In this
+initial proposal, the Trigger must explicitly enable dynamic parsing by setting
+boolean fields `parseExtensions` and/or `parseData`. Future work may
+eliminate the need for these fields by detecting the use of extension attributes
+and/or data fields in the expression.
+
+### Why CEL?
+
+CEL is the
+[Common Expression Language](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/doc/langdef.md).
+It is designed for expressing security policy and protocols, and has the
+following
+[guiding philosophy](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/README.md):
+
+*   **Small and fast.** CEL is not Turing complete, so it is easy to understand
+    and implement.
+*   **Extensible.** CEL is designed to be embedded and can be extended with new
+    functionality.
+*   **Developer-friendly.** The language spec was based on experience and
+    usability testing from implementing Firebase Rules.
+
+Additional properties of CEL that make it attractive for Trigger Filtering
+include:
+
+*   **Familiar expression syntax.** Filter expressions written in CEL look and
+    behave like conditional expressions written in the most popular languages.
+*   **Expressions can be compiled and cached for efficiency.** We expect users
+    will create many filters that change rarely in comparison to frequency of
+    evaluations. This usage pattern works in favor of an expression language
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+```
+
+Specified with a CEL expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
```

Comment:
trying to think of an alternative, but geez that's a lot of layers just for a simple/single `==`.

Reply:
The purpose of the `cel` level to allow for additional fields that are relevant only to the CEL filter strategy, like `parseData` and `parseExtensions`. I agree, it's a lot of layers. I think this can be worked around by offering a structured syntax (like the current `SourceAndType`) that can express most filter cases, so that most users will only use CEL when their needs are complex enough that they don't worry about the verbosity of the syntax.

If you come up with something more elegant, I'd love to see it @duglin :)